lib/datadog/appsec/contrib/rails/gateway/watcher.rb

```diff
@@ -0,0 +1,81 @@
+# typed: false
+
+require 'datadog/appsec/instrumentation/gateway'
+require 'datadog/appsec/reactive/operation'
+require 'datadog/appsec/contrib/rails/reactive/action'
+require 'datadog/appsec/event'
+
+module Datadog
+  module AppSec
+    module Contrib
+      module Rails
+        module Gateway
+          # Watcher for Rails gateway events
+          module Watcher
+            def self.watch
+              Instrumentation.gateway.watch('rails.request.action') do |stack, request|
+                block = false
+                event = nil
+                waf_context = request.env['datadog.waf.context']
+
+                AppSec::Reactive::Operation.new('rails.request.action') do |op|
+                  trace = active_trace
+                  span = active_span
+
+                  Rails::Reactive::Action.subscribe(op, waf_context) do |action, result, _block|
+                    record = [:block, :monitor].include?(action)
+                    if record
+                      # TODO: should this hash be an Event instance instead?
+                      event = {
+                        waf_result: result,
+                        trace: trace,
+                        span: span,
+                        request: request,
+                        action: action
+                      }
+
+                      waf_context.events << event
+                    end
+                  end
+
+                  _action, _result, block = Rails::Reactive::Action.publish(op, request)
+                end
+
+                next [nil, [[:block, event]]] if block
```

Comment:
I find this kind of return value hard to reason about, something like:
```ruby
next FooResult.new(result: nil, action: :block, event: event) if block
```
would be much easier to follow IMO.

Not for this PR, but something I think would help legibility here.

Reply:
Agreed: same reasoning as `Event` over there!